src/Symfony/Bundle/FrameworkBundle/Tests/Test/WebTestCaseTest.php

```diff
@@ -13,340 +13,469 @@
+
+        $this->assertResponseNotHasHeader('X-Date');
+
         $this->expectException(AssertionFailedError::class);
         $this->expectExceptionMessage('Failed asserting that the Response does not have header "Date".');
-        $this->getResponseTester(new Response())->assertResponseNotHasHeader('Date');
+
+        $this->assertResponseNotHasHeader('Date');
     }
 
     public function testAssertResponseHeaderSame()
     {
-        $this->getResponseTester(new Response())->assertResponseHeaderSame('Cache-Control', 'no-cache, private');
+        $client = static::createClient();
+        $client->request('GET', '/200');
+
+        $this->assertResponseHeaderSame('Cache-Control', 'no-cache, private');
+
         $this->expectException(AssertionFailedError::class);
         $this->expectExceptionMessage('Failed asserting that the Response has header "Cache-Control" with value "public".');
-        $this->getResponseTester(new Response())->assertResponseHeaderSame('Cache-Control', 'public');
+
+        $this->assertResponseHeaderSame('Cache-Control', 'public');
     }
 
     public function testAssertResponseHeaderNotSame()
     {
-        $this->getResponseTester(new Response())->assertResponseHeaderNotSame('Cache-Control', 'public');
+        $client = static::createClient();
+        $client->request('GET', '/200');
+
+        $this->assertResponseHeaderNotSame('Cache-Control', 'public');
+
         $this->expectException(AssertionFailedError::class);
         $this->expectExceptionMessage('Failed asserting that the Response does not have header "Cache-Control" with value "no-cache, private".');
-        $this->getResponseTester(new Response())->assertResponseHeaderNotSame('Cache-Control', 'no-cache, private');
+
+        $this->assertResponseHeaderNotSame('Cache-Control', 'no-cache, private');
     }
 
     public function testAssertResponseHasCookie()
     {
-        $response = new Response();
-        $response->headers->setCookie(HttpFoundationCookie::create('foo', 'bar'));
+        $client = static::createClient();
+        $client->request('GET', '/200');
+
+        $this->assertResponseHasCookie('foo');
 
-        $this->getResponseTester($response)->assertResponseHasCookie('foo');
         $this->expectException(AssertionFailedError::class);
         $this->expectExceptionMessage('Failed asserting that the Response has cookie "bar".');
-        $this->getResponseTester($response)->assertResponseHasCookie('bar');
+
+        $this->assertResponseHasCookie('bar');
     }
 
     public function testAssertResponseNotHasCookie()
     {
-        $response = new Response();
-        $response->headers->setCookie(HttpFoundationCookie::create('foo', 'bar'));
+        $client = static::createClient();
+        $client->request('GET', '/200');
+
+        $this->assertResponseNotHasCookie('bar');
 
-        $this->getResponseTester($response)->assertResponseNotHasCookie('bar');
         $this->expectException(AssertionFailedError::class);
```

Comment:
AFAIK, expecting the PHPUnit failures is also deprecated. So I don't think this is the right approach.

Your new test is testing WebTestCaseController. It is not actually writing tests for WebTestCase itself (it only test it as a side effect).

Actually asserting the behavior of a custom base TestCase properly is more complex. You could look at https://github.com/phpspec/prophecy-phpunit where I did such things

Reply:
I see your points. Let's try another approach to reach compatibility with PHPUnit 10+ (see #58517)